modulemd/v2/modulemd-validator.c

```diff
@@ -0,0 +1,213 @@
+  return TRUE;
+}
+
+// clang-format off
+static GOptionEntry entries[] = {
+  { "quiet", 'q', G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK, set_verbosity, "Print no output", NULL },
+  { "verbose", 'v', G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK, set_verbosity, "Be verbose", NULL },
+  { "debug", 0, G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK, set_verbosity, "Output debugging messages", NULL },
+  { G_OPTION_REMAINING, 0, 0, G_OPTION_ARG_FILENAME_ARRAY, &options.filenames, "Files to be validated", NULL },
+  { NULL } };
+// clang-format on
+
+
+static gboolean
+parse_file (const gchar *filename, GPtrArray **failures, GError **error)
+{
+  MMD_INIT_YAML_PARSER (parser);
+  MMD_INIT_YAML_EVENT (event);
+  g_autoptr (FILE) yaml_stream = NULL;
+  int saved_errno;
+  g_autoptr (ModulemdModuleIndex) index = NULL;
+
+  yaml_parser_initialize (&parser);
+
+  if (options.verbosity >= MMD_VERBOSE)
+    {
+      g_fprintf (stdout, "Validating %s\n", filename);
+    }
+
+  /* Parse documents */
+  yaml_stream = g_fopen (filename, "rb");
+  saved_errno = errno;
+
+  if (yaml_stream == NULL)
+    {
+      if (options.verbosity >= MMD_DEFAULT)
+        {
+          g_fprintf (stdout,
+                     "Failed to open file %s: %s\n",
+                     filename,
+                     g_strerror (saved_errno));
+        }
+      return FALSE;
+    }
+
+
+  yaml_parser_set_input_file (&parser, yaml_stream);
+
+  index = modulemd_module_index_new ();
+  return modulemd_module_index_update_from_parser (
+    index, &parser, TRUE, TRUE, failures, error);
+}
+
+
+int
+main (int argc, char *argv[])
+{
+  const char *filename;
+  GOptionContext *context;
+  g_autoptr (GError) error = NULL;
+  gsize num_invalid = 0;
+  gboolean ret;
+  g_autoptr (GPtrArray) failures = NULL;
+  ModulemdSubdocumentInfo *doc = NULL;
+
+  setlocale (LC_ALL, "");
+
+  context = g_option_context_new ("FILES - Simple modulemd YAML validator");
+  g_option_context_add_main_entries (context, entries, "modulemd-validator");
+  if (!g_option_context_parse (context, &argc, &argv, &error))
+    {
+      g_print ("option parsing failed: %s\n", error->message);
+      exit (1);
+    }
+
+  if (!(options.filenames && options.filenames[0]))
+    {
+      g_fprintf (stderr,
+                 "At least one file must be specified on the command-line\n");
+      return EXIT_FAILURE;
+    }
+
+  for (gsize i = 0; options.filenames[i]; i++)
+    {
+      filename = options.filenames[i];
+      failures = g_ptr_array_new_with_free_func (g_object_unref);
+
+      ret = parse_file (filename, &failures, &error);
+      if (!ret)
+        {
+          num_invalid++;
+          if (options.verbosity >= MMD_DEFAULT)
+            {
+              g_fprintf (stderr, "%s failed to validate\n", filename);
+
+              if (error != NULL)
+                {
+                  /* Unparseable content */
+                  g_fprintf (
+                    stderr, "%s contained unparseable YAML\n", filename);
```

Comment:
Maybe also print the `error`?

Reply:
I probably should, but the error is going to be "Parse Error" pretty much 100% of the time. I suppose I should be future-compatible in case it becomes possible for a different error to come back.